packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -4157,3 +4164,244 @@ class _CopySelectionAction extends ContextAction<CopySelectionTextIntent> {
   @override
   bool get isActionEnabled => state._value.selection.isValid && !state._value.selection.isCollapsed;
 }
+
+/// A void function that takes a [TextEditingValue].
+@visibleForTesting
+typedef TextEditingValueCallback = void Function(TextEditingValue value);
+
+/// Provides undo/redo capabilities for text editing.
+///
+/// Listens to [controller] as a [ValueNotifier] and saves relevant values for
+/// undoing/redoing. The cadence at which values are saved is a best
+/// approximation of the native behaviors of a hardware keyboard on Flutter's
+/// desktop platforms, as there are subtle differences between each of these
+/// platforms.
+///
+/// Listens to keyboard undo/redo shortcuts and calls [onChanged] when a
+/// shortcut is triggered that would affect the state of the [controller].
+@visibleForTesting
+class TextEditingHistory extends StatefulWidget {
```

Comment:
Can we keep this private?

Reply:
Yes good catch, I'm no longer using this directly in the tests, so it can be fully private.